rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,99 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+serialized at HEAD with existing [`stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc)
+test files.
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing
+`stablehlo_legalize_to_vhlo.mlir` at head and comparing it to a known-good file
+serialized serialized at previous releases. This approach operates on the
+assumption that at HEAD we are able to produce a byte-identical serialized
+artifact, excluding custom header info and debug locations.
```

Comment:
"excluding producer information and locations"? 1) I don't think "custom header" is a thing in [the bytecode format spec](https://mlir.llvm.org/docs/BytecodeFormat/). 2) Also, locations are called just "locations", not "debug locations".

Reply:
Will do. The current wording is based on `--mlir-pretty-debuginfo` / `--strip-debuginfo`, which prints / strips locations (and only locations, looked at the strip debuginfo pass).

Reply:
That's just for the textual format, those flags don't affect bytecode. That being said, locations should be retained, for the .mlir file you may want to explicitly specify all locations so that it is roundtrip.

Reply:
I believe `strip-debuginfo` runs a pass on the IR to replace locations with `Unknown`?

We could include them and standardize the directory that tests are run from, that may not work well when StableHLO gets exported to other projects. We could also consider explicitly specifying locations, but that is a bit cumbersome in the test file (2.2k line test file).

Reply:
Created https://github.com/openxla/stablehlo/issues/1633 to experiment with keeping source locations in these test files.